CIP-xxxxx/README.md

```diff
@@ -0,0 +1,218 @@
+
+```ts
+type KeyHash = String
+```
+
+#### ScriptRequirements
+
+Script requirements encapsulate all the possible requirements for a transaction to be valid. It includes the following fields:
+
+- collateral: The list of inputs that will be used as collateral for the transaction.
+- inputs: The list of inputs that will be used as inputs for the transaction.
+- reference_inputs: The list of inputs that will be used as reference inputs for the transaction.
+- outputs: The list of outputs that will be used for the transaction.
+- mint: The amount of tokens that will be minted/burned in the transaction.
+- certificates: The list of certificates that will be included in the transaction.
+- withdrawals: The list of withdrawals that will be included in the transaction.
+- validity_range: The validity range that the transaction must be valid in.
+- signatories: The list of signatories that must sign the transaction.
+- redeemers: The list of redeemers that will be used in the transaction.
+- datums: The list of datums that will be used in the transaction.
+
+
+```ts
+type ScriptRequirement = {
+  collateral?: List<cbor<transaction_unspent_output>>,
+  inputs?: List<cbor<transaction_unspent_output>>,
+  reference_inputs?: List<cbor<transaction_unspent_output>>,
+  outputs?: List<transaction_output>,
+  mint?: Value,
+  certificates?: List<Certificate>,
+  withdrawals?: Dict<StakeCredential, Int>,
+  validity_range?: ValidityRange,
+  signatories?: List<KeyHash>,
+  redeemers?: Dict<ScriptPurpose, Redeemer>,
+  datums?: Dict<Hash<Blake2b_256, Data>, Data>
+}
+```
+
+#### ValidityRange
+
+```ts
+type ValidityRange = {
+  valid_before?: PosixTime,
+  valid_after?: PosixTime,
+}
+```
+
+#### Value
+```ts
+type Value = {
+    amount: Coin,
+    asset: String
+}
+```
+
+### Aditional Error Types
+
+#### CompletedTxError
+
+```ts
+CompletedTxErrorCode = {
+	NotFound: 1,
+	NotReady: 2
+}
+```
+
+* NotFound - The transaction with the given id was not found.
+* NotReady - The transaction with the given id is not ready yet. 
+
+## Motivation: why is this CIP necessary?
+
+In order to facilitate future dApp development, we will need a way for dApps to communicate with Plutus wallets. Given the unique complexities of Plutus script-based addresses, special provisions need to be made to make the connector compatible with them.
+
+Specifically, apps building transactions need to be able to get the following information from the wallet:
+
+- Script descriptor
+- Script Requirements list
+- Collateral donator
+
+Additionally, apps need to be able to submit a transaction to the wallet for signing in an asynchronous manner, as gathering of signatures can take a long time and each wallet provider will have its own way of handling this process.
+
+Finally, the `signTx()` and `signData()` endpoints will have to be disabled when using this extension since they are not compatible with native-script-based addresses.
+
+## Rationale for the required data
```

Comment:
@leo42 I think I see why you've written it in this order (to keep the rationale close to the data structure it refers to) but I still think you should move it down into the usual _Rationale_ section which currently contains nothing but a reference back here.

Remember also you state that this document is a "work in progress" which means this rationale might not always apply, nor apply to future versions.  Therefore as each section grows it will help us (editors, implementors, and future version authors) to maintain the standard document organisation without forward & backward references.

Reply:
resolving: moved to proper _Rationale_ section as suggested.